### PR TITLE
Add emacs-lisp-check-declare checker

### PR DESCRIPTION
**Pull request by darkfeline**

### CODE REVIEW

**Comment by darkfeline** [comment]
Hm, this appears to work when I load the checker into my working Emacs, but the flycheck-test test doesn't pass, and ert is woefully opaque about why.

**Comment by fmdkdd** [comment]
The output to ERT is not the most intuitive:
```
((should
       (equal expected flycheck-current-errors))
      :form
      (equal
       ([cl-struct-flycheck-error #<killed buffer> emacs-lisp-check-declare "/home/travis/build/flycheck/flycheck/test/resources/language/emacs-lisp/check-declare-errors.el" 10 1 "said ‘this-function-is-not-defined’ was defined in this-file-does-not-exist.el: file not found" warning nil nil])
       nil)
```
Inside the `:form (equal`, it expects the error you declared, but you got `nil` instead.  That means the error was either not caught by the checker, or the checker was not run at all.

I think the issue here is that your checker is never *selected* by the Flycheck run by ERT.  `emacs-lisp` is the default checker, followed by `emacs-lisp-checkdoc` (see the `:next-checker` property).

You could put your checker as a `:next-checker` for `checkdoc`.

**Comment by darkfeline** [comment]
Ah, got it.  The tests pass now.

File: test/resources/language/emacs-lisp/check-declare-errors.el [hunk 1]
```diff
@@ -0,0 +1,13 @@
+;;; check-declare-errors.el -- check-declare errors -*- lexical-binding: t; -*-
+
+;;; Commentary:
+
+;; Checkdoc uses the syntax table and comment syntax of the current buffer; if
```
Comment (cpitclaudel) [inline test/resources/language/emacs-lisp/check-declare-errors.el]: Isn't that comment a leftover from another file?

File: flycheck.el [hunk 2]
```diff
@@ -7161,12 +7162,34 @@ The checker runs `checkdoc-current-buffer'."
     (not (or (flycheck-autoloads-file-p)
              (and (buffer-file-name)
                   (member (file-name-nondirectory (buffer-file-name))
-                          '("Cask" "Carton" ".dir-locals.el")))))))
+                          '("Cask" "Carton" ".dir-locals.el"))))))
+  :next-checkers (emacs-lisp-check-declare))
```
Comment (cpitclaudel) [inline flycheck.el]: Shouldn't this be added to the next-checkers of emacs-lisp as well?

**Comment by fmdkdd** [comment]
>  shouldn't this be part of the default ELisp checker?

I agree.  It looks like we could just add the `check-declare-form` as an additional argument to the `emacs-lisp` checker, and save ourselves the trouble of calling `emacs` a third time.

**Comment by fmdkdd** [comment]
(Reminder: #1229)

**Comment by darkfeline** [comment]
Merged checker into emacs-lisp

**Comment by darkfeline** [comment]
Looks like I need to get my hands on a build of Emacs 24, please wait warmly.

**Comment by darkfeline** [comment]
Actually, it looks like I can't build Emacs 24.  I believe I'm running into the bug mentioned here: http://lists.linuxfromscratch.org/pipermail/blfs-support/2016-August/078284.html

It may be a while before I can get Emacs 24.  If someone else can fix this, feel free.  I suspect it's something like the error filter needing to be tweaked.

**Comment by darkfeline** [comment]
The Emacs I used was 24.3, which flycheck doesn't support.  If the fix doesn't pass, I'll probably install a newer Ubuntu in a VM and try again.

I wonder if flycheck would benefit from shipping a Vagrantfile.

**Comment by darkfeline** [comment]
Well, that was a waste of time.  As it turns out, check-declare does not print the line number on Emacs 24 and stepping through the debugger I learned that Flycheck discards errors that do not have line numbers.

I don't really feel like trying to make this work now, so I've skipped the test on Emacs 24 and documented that check-declare doesn't work on Emacs 24.

**Comment by fmdkdd** [comment]
Thanks for investigating the issue.  So, currently the `check-declare` pass will be run by Flycheck on Emacs 24, but the errors won't be reported, is that it?

If the issue is missing line numbers in the output, you can check for errors that have a `nil` line and set them to 0 instead.  That way, the errors will be reported on the first line, and Flycheck will not discard them.

If it's too much a hassle to test that with Emacs 24, let me know.  I have an Emacs 24 install I can use to test this workaround.

**Comment by darkfeline** [comment]
Currently on Emacs 24, `check-declare` will emit errors correctly, but they will be filtered out by `flycheck-relevant-errors`/`flycheck-relevant-error-p`.

Adding 0 line numbers to them sounds like a good solution; I'll try to do it with error-filters.

**Comment by darkfeline** [comment]
Any update on this?

File: flycheck.el [hunk 3]
```diff
@@ -6946,7 +6959,10 @@ See URL `https://github.com/lpil/dogma/'."
       (unwind-protect
           (byte-compile-file (car command-line-args-left))
         (mapc (lambda (f) (ignore-errors (delete-file f)))
-              flycheck-byte-compiled-files)))))
+              flycheck-byte-compiled-files))
+      (when (and (boundp 'flycheck-emacs-lisp-check-declare)
+                 flycheck-emacs-lisp-check-declare)
```
Comment (cpitclaudel) [inline flycheck.el]: Could this be written as `bound-and-true-p`?

File: flycheck.el [hunk 4]
```diff
@@ -3617,6 +3617,19 @@ IDs, that include the name of the tool."
...
+  "Set ERRORS without lines to line 0.
+
+Use as `:error-filter' for syntax checkers that output errors
+without line numbers.
+
+Return ERRORS."
+  (seq-do (lambda (err)
+            (unless (flycheck-error-line err)
+              (setf (flycheck-error-line err) 0)))
+          errors)
```
Comment (cpitclaudel) [inline flycheck.el]: For simple iteration like this, I tend to prefer dolist over seq-do

File: flycheck.el (hunk 3 above)
Comment (darkfeline) [inline flycheck.el]: Done

File: flycheck.el (hunk 4 above)
Comment (darkfeline) [inline flycheck.el]: All of the other filters use `seq-do`, so I am inclined to keep the code consistent.

**Comment by cpitclaudel** [comment]
Brilliant work, thanks!

**Comment by fmdkdd** [comment]
Thank you @darkfeline for your patience, and for this nice contribution!

**Pull request merged**